src/lib/sql/sql_translator.cpp

```diff
@@ -403,15 +410,103 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_join(const hsql::Join
                                                    : std::make_pair(*left_in_right_node, *right_in_left_node);
 
   // Joins currently only support one simple condition (i.e., not multiple conditions).
-  auto predicate_condition = translate_operator_type_to_predicate_condition(condition.opType);
-
+  auto predicate_condition = translate_operator_type_to_predicate_condition(join_condition->opType);
   auto join_node = JoinNode::make(join_mode, column_references, predicate_condition);
   join_node->set_left_input(left_node);
   join_node->set_right_input(right_node);
-
+  _apply_predicate_expressions(join_node, LQPInputSide::Left, left);
+  _apply_predicate_expressions(join_node, LQPInputSide::Right, right);
   return join_node;
 }
 
+LQPInputSide SQLTranslator::_get_side(const std::shared_ptr<AbstractLQPNode>& left_node,
+                                      const std::shared_ptr<AbstractLQPNode>& right_node, const hsql::Expr* expr) {
+  DebugAssert(expr && expr->type == hsql::kExprColumnRef, "only applicable to columnar expressions");
+  const auto qualified = HSQLExprTranslator::to_qualified_column_name(*expr);
+  if (left_node->find_column(qualified)) {
+    return LQPInputSide::Left;
+  }
+
+  if (right_node->find_column(qualified)) {
+    return LQPInputSide::Right;
+  }
+
+  Fail("Column not found: " + qualified.as_string());
+}
+
+void SQLTranslator::_get_sides_from_operator_expressions(std::vector<const hsql::Expr*> operators,
+                                                         std::function<LQPInputSide(const hsql::Expr*)> get_side,
+                                                         std::vector<const hsql::Expr*>& both,
+                                                         std::vector<const hsql::Expr*>& left,
+                                                         std::vector<const hsql::Expr*>& right) {
+  for (const auto* op : operators) {
+    const bool left_is_column = op->expr && op->expr->type == hsql::kExprColumnRef;
+    const bool right_is_column = op->expr2 && op->expr2->type == hsql::kExprColumnRef;
+
+    if (left_is_column && right_is_column) {
```

Comment:
I think you can have it shorter:

```
    if (left_is_column && right_is_column && get_side(op->expr) != get_side(op->expr2)) {
      both.push_back(op);
      continue;
    }
    if (left_is_column) {
      auto& side = (get_side(op->expr) == LQPInputSide::Left) ? left : right;
      side.push_back(op);
    } else if (right_is_column) {
      auto& side = (get_side(op->expr2) == LQPInputSide::Left) ? left : right;
      side.push_back(op);
    } else {
      // neither is a column - not sure what you would do here
    }
```

Reply:
Thanks, this is what I really hoped to get done during review. Previously it was easy to understand but very verbose.
I think the case where two columns are from the same side (`tableA.c1 = tableA.c2`) is handled correctly in above proposal, but I really would like to make it explicit - else this case is too easy to overlook: currently it is only working correctly because the left & right column branch are mutually exclusive.

Reply:
For me, that would be explicit enough. If they are from the same side, they are not from opposite sides (which is handled in the first line). So they can be either from the left or from the right.